fan: Add tachometer support

arkeet opened the pull request:
In response to https://github.com/KevinOConnor/klipper/issues/3771

This adds two new config options for fans: 'tachometer_pin' to specify
the GPIO pin, and 'min_rpm' to ensure that the fan is spinning fast
enough (and shutdown otherwise). The rpm is also exposed in get_status
for command templates and the api server.

There is a new mcu object for a generic edge counter, which repeatedly
polls a GPIO pin and periodically reports the count to the host.

arkeet commented:
Some architectural questions: 
- Should I move the frequency calculation into the counter class, and pass it through the callback? Then the fan code doesn't need to care about the absolute number of counts.
- Should the rpm monitoring code be moved to its own class? This may make it cleaner especially since in the future I think it would be good to support always-on fans (with no mcu output pin), as well as fans that have a simple stall sensor output (that just pulls constantly low while the fan is spinning).

KevinOConnor commented:
Interesting - thanks.  My high-level feedback is that the functionality looks useful.

I noticed a couple of things that I think should probably be addressed prior to merging:
1. I think it would be preferable to merge the base tach support first, and then add the "min_rpm" support in a later PR.  I'm a little leery of invoking a shutdown with new code and wonder if it would be better to get reporting merged first.
2. I do think it would be better to move the logic into its own class.  Since there is already a new counter.py file, perhaps add a counter.py:TachCounter() class that does all the work.  I'd guess the fan code could just have `self.tach = counter.TachCounter(config)` in its init, and `return dict(self.tach.get_status(eventtime), speed=self.lst_fan_value)` in its get_status() method.
3. I think the name counter.py is a little too generic.  Maybe something like pulse_counter.py
4. I'm a little leery of using "rpm" in the config, as introducing a unit in minutes can be very confusing.  In general, I'd prefer all config items that use time to use a unit in seconds.

Some other minor notes:
1. The `counter_event()` c code can be optimized.  It's not necessary to directly call `timer_read_time()` (one can get an estimate via `c->timer.waketime`).  It looks like `c->next_poll_time` is always the same as c->timer.waketime and could thus be removed.  It's possible to remove `c->value` by using `c->flags` - for example `if ((value ^ c->flags) & CF_LAST_VALUE) { ... }`.  It's not necessary to store `c->last_count_time` as the counter_state message can use `c->timer.waketime` (an example of that is in MCU_adc).
2. In other messages you raised the possibility of using hardware irqs to optimize the code.   FWIW, I don't think that will improve performance sufficiently to be worthwhile.  If you did wish to handle very high update rates, I'd guess it's best to go directly to support for "hardware timer/counter units".  Though, the basic software support is probably fine.

-Kevin

mattthebaker commented:
Cool feature, I'd like to see tachometer feedback support.

Wanted to throw in a quick example with numbers to get a handle on timing requirements.

Here is a tachometer feedback blower: https://www.digikey.com/en/products/detail/delta-electronics/BFB0405HHA-TZP4/3307853 .  It runs at 8500 rpm +5/-15%, and has 8 poles. So it will pulse 8500*1.05*8/60 = 1190 times per second. In order to avoid aliasing in the measurement, the pin will need to be sampled at least at the Nyquist frequency, or 2*1190 = 2380 Hz. Its also useful to think about the inverse, the sampling interval, 1s/2380 = 420us. If the firmware cannot guarantee it will check the pin that often under the worst case conditions, the measurement may be impacted. Without having exhaustively searched for fans, at least 5-10kHz may be necessary.

Hope that can help inform the decision whether or not hardware counter peripherals are necessary for tachometer pin monitoring.

arkeet commented:
@KevinOConnor, thanks for the feedback. I'll hopefully get some time to revise this in the next couple days. I'll remove the min_rpm stuff for now, and leave it for a future PR.

> 4. I'm a little leery of using "rpm" in the config, as introducing a unit in minutes can be very confusing.  In general, I'd prefer all config items that use time to use a unit in seconds.

I understand the desire for consistency, but fan rotation speeds are, as far as I've seen, always specified in RPM, so I think it would be much more user friendly for comparing readings with the spec sheet or choosing a `min_rpm` value.

> It's not necessary to store `c->last_count_time` as the counter_state message can use `c->timer.waketime` (an example of that is in MCU_adc).

The `last_count_time` is different from `waketime` because it is the last the `count` was incremented, which is in general earlier than the time the sample is reported back to the host. This is intentional, to improve resolution reduce jitter in computed rpm values, since we can divide the number of edges by the actual time interval between them, rather than the fixed report interval (which is in general not aligned with the edges).

---

> Here is a tachometer feedback blower:

@mattthebaker, thanks for the info. That fan has twice as many poles as typical fans like the ones I've looked at, so it seems we would need another config option like `tachometer_ppr` specify the number of pulses (= full square wave cycles) per revolution; we can use a default value of 2, but that fan would need 4. 

By the way, the max signal frequency (full waves per second) on that fan is just 595 Hz, so the Nyquist frequency is 1190 Hz. I think a 5 kHz polling interval would be fast enough - that could handle 8-pole fans up to 37,500 RPM, or 4-pole fans up to 75,000 RPM. (For reference I only see one BLDC fan on Digi-Key with tach output over 37,500 RPM - and it's a 4-pole fan.) @KevinOConnor do you see any issues with such a fast polling rate? Should it be configurable somehow?

mattthebaker commented:
@arkeet Ahh, yeah, you're right that fan is 4ppr not 8ppr. Agreed that the config would need a ppr configuration option. 2-5kHz polling is encouraging for the general case vs 5-10kHz.

KevinOConnor commented:
> do you see any issues with such a fast polling rate? Should it be configurable somehow?

It's not easy to provide a polling maximum - it'll be dependent on the type of micro-controller and the other tasks the mcu needs to accomplish.  I'd guess that the number of times the mcu can poll will be similar to the number of steps per second the mcu can execute - as both are really tests of the Klipper mcu scheduler.  See: https://www.klipper3d.org/Features.html#step-benchmarks .  Will a user be okay with giving up 10% of their avr mcu for a fan?  40% for 4 fans?  I don't know.

If you wanted to, it should be possible to confirm the impact of tach polling by running a variant of the step rate benchmark described at: https://www.klipper3d.org/Benchmarks.html

Allowing the polling time to be configured sounds like a good idea.  Along with a conservative default.

Cheers,
-Kevin

xkuyax commented:
> @KevinOConnor, thanks for the feedback. I'll hopefully get some time to revise this in the next couple days. I'll remove the min_rpm stuff for now, and leave it for a future PR.
> 
> > 1. I'm a little leery of using "rpm" in the config, as introducing a unit in minutes can be very confusing.  In general, I'd prefer all config items that use time to use a unit in seconds.
> 
> I understand the desire for consistency, but fan rotation speeds are, as far as I've seen, always specified in RPM, so I think it would be much more user friendly for comparing readings with the spec sheet or choosing a `min_rpm` value.
> 
> > It's not necessary to store `c->last_count_time` as the counter_state message can use `c->timer.waketime` (an example of that is in MCU_adc).
> 
> The `last_count_time` is different from `waketime` because it is the last the `count` was incremented, which is in general earlier than the time the sample is reported back to the host. This is intentional, to improve resolution reduce jitter in computed rpm values, since we can divide the number of edges by the actual time interval between them, rather than the fixed report interval (which is in general not aligned with the edges).
> 
> > Here is a tachometer feedback blower:
> 
> @mattthebaker, thanks for the info. That fan has twice as many poles as typical fans like the ones I've looked at, so it seems we would need another config option like `tachometer_ppr` specify the number of pulses (= full square wave cycles) per revolution; we can use a default value of 2, but that fan would need 4.
> 
> By the way, the max signal frequency (full waves per second) on that fan is just 595 Hz, so the Nyquist frequency is 1190 Hz. I think a 5 kHz polling interval would be fast enough - that could handle 8-pole fans up to 37,500 RPM, or 4-pole fans up to 75,000 RPM. (For reference I only see one BLDC fan on Digi-Key with tach output over 37,500 RPM - and it's a 4-pole fan.) @KevinOConnor do you see any issues with such a fast polling rate? Should it be configurable somehow?

Do you think we would really need to measure the full 37k RPM? For me 1/10th of that is enough to make sure my fan is spinning

mattthebaker commented:
@xkuyax unfortunately, yes, you do. This is the nature of aliasing and the reason for sampling at the nyquist frequency.

Lets say you decide to sample the tachometer at 500Hz. If your fan tach pulses at exactly 500Hz, 1000Hz, 1500Hz, it may appear that your fan is not moving at all. Every time you look at the tach input, it could be in the same state. This is aliasing. When the fan is spinning very close to a multiple of the sampling rate, it will look like it is spinning very slow. A prime example of this effect is videos of cars in motion, at certain speeds the wheels appear to spin very slowly or not at all. Basically, you need to sample at least 2x the fastest the tach is expected to toggle to measure it correctly, that is the nyquist frequency.

For slower fans, the sampling rate can be reduced proportionally. Once the framework is in place it could also be possible to use counter hardware built in many controllers instead of polling.

arkeet commented:
Updated based on Kevin's feedback.
* Everything related to `min_rpm` is removed and shall be in a future PR.
* Reported speed is still in units of RPM.
* New `FrequencyCounter` class handles the logic of turning counts from the mcu into frequency values. Instead of this providing `get_status` the fan code provides a callback which does the trivial conversion from pulse frequency to RPM.
* C changes: `next_poll_time` is gone; `last_count_time` is still there for reasons I described previously; `value` and `flags` have been turned into bitfields.
* Changed the polling frequency to 4 kHz - enough for 60k RPM fans but still should be light enough? (< 3% of the step rate on a 16MHz AVR according to the [step benchmarks](https://www.klipper3d.org/Features.html#step-benchmarks)) I haven't run any benchmarks myself. Didn't add a config option for this (don't see much benefit and would may be confusing to document), but I'm open for suggestions.

KevinOConnor commented on klippy/extras/fan.py:
```diff
@@ -28,9 +30,22 @@ def __init__(self, config, default_shutdown_speed=0.):
         self.mcu_fan.setup_start_value(0., shutdown_power)
+
+        # Setup tachometer
+        tach_pin = config.get('tachometer_pin', None)
+        if tach_pin is not None:
+            self.tach_ppr = config.getint('tachometer_ppr', 2, minval=1)
+            POLL_TIME = 1. / 4000 # enough for 60000 rpm max at 2 ppr
+            SAMPLE_TIME = 1.
+            tach_counter = pulse_counter.FrequencyCounter(
+                self.printer, tach_pin, SAMPLE_TIME, POLL_TIME)
```
FWIW, I think it would be worthwhile to add a `config.getfloat('tachometer_poll_interval`, .002)` config option and default it to a conservative (not very high) value.  I'd guess most fans aren't super fast and I would guess most users would prefer a conservative value that definitely wont impact their stepping rates.

KevinOConnor commented on klippy/extras/fan.py:
```diff
@@ -54,8 +69,15 @@ def set_speed_from_command(self, value):
         self.set_speed(print_time, 0.)
+
+    def tach_callback(self, time, freq):
+        self.rpm = freq * 30. / self.tach_ppr # 2 ppr encoder
+
     def get_status(self, eventtime):
-        return {'speed': self.last_fan_value}
+        return {
+            'speed': self.last_fan_value,
+            'rpm': self.rpm,
```
FWIW, I think it would be preferable to try and avoid the use of callbacks here.  The callback is run in a background thread, and many developers may not realize that.  I think it would be preferable if the main class that calls `mcu.register_response()` be the only class to deal with the multi-threading issues.  (That is, it can obtain the values, store them internally, and make them available on request.  Ideally, that would allow the calculations to be done entirely from the main thread and limit the chance of multi-threading issues.

KevinOConnor commented on src/counter.c:
```diff
@@ -0,0 +1,98 @@
+        c->last_count_time = time;
+    }
+
+    if (time - c->last_sample_time >= c->sample_ticks) {
```
This should use `timer_is_before()`.  It's preferable to use that call instead of direct comparisons so as to avoid any issues with counter rollover.

KevinOConnor commented on src/counter.c:
```diff
@@ -0,0 +1,98 @@
+    uint32_t poll_ticks;
+    uint32_t sample_ticks, last_sample_time;
+    uint32_t count, last_count_time;
+    bool value          : 1;
+    bool sample_pending : 1;
```
FWIW, the use of C bitmap fields should be avoided.  Their behaviour is not well defined by the C specifications, and they can cause subtle issues depending on how the compiler implements them.  Indeed, they may have different implementations on AVR vs ARM vs PRU, etc.

So, as an example, were `sample_pending` to be updated outside of an `irq_enable(); ... ; irq_disable();` block, it could subtly add a read-modify-write operation to `value` and possibly introduce a race-condition.  I don't see anything in your code that could cause a problem, but I fear it's a "land mine" for future developers.  IMO, best to avoid C bitmaps.  Being explicit with flag manipulation is only a couple of extra lines of code and doesn't have these subtle issues.

KevinOConnor commented on src/Makefile:
```diff
@@ -7,4 +7,4 @@ src-$(CONFIG_HAVE_GPIO_SPI) += spicmds.c thermocouple.c
 src-$(CONFIG_HAVE_GPIO_HARD_PWM) += pwmcmds.c
 src-$(CONFIG_HAVE_GPIO_BITBANGING) += lcd_st7920.c lcd_hd44780.c buttons.c \
-    tmcuart.c spi_software.c neopixel.c sensor_adxl345.c
+    tmcuart.c spi_software.c neopixel.c sensor_adxl345.c counter.c
```
FYI, I think counter.c may be too generic a name.  Perhaps `sensor_counter.c`?

KevinOConnor commented on src/counter.c:
```diff
@@ -0,0 +1,98 @@
+        if (!c->sample_pending)
+            continue;
+        irq_disable();
+        uint32_t time = c->last_sample_time;
+        uint32_t count = c->count;
+        uint32_t count_time = c->last_count_time;
+        c->sample_pending = false;
+        irq_enable();
+        sendf("counter_state oid=%c time=%u count=%u count_time=%u",
+              oid, time, count, count_time);
```
Wouldn't it be more accurate to report `c->time.waketime - c->poll_ticks` instead of `c->last_sample_time` - the `count` may have changed between the `sched_wake_task()` call and the start of this code.

KevinOConnor reviewed (commented):
Thanks.  At a high-level it looks good to me.  I have a few minor comments below.

-Kevin

arkeet commented:
Hi Kevin, sorry for taking so long with this. I believe I've addressed all your concerns. I put all the fan tachometer stuff in a `FanTachometer` class (which handles the config parsing and rpm conversion). There is still a callback from `MCU_counter` to `FrequencyCounter` - I wanted to keep the frequency counting logic separate, and I think changing it to not use callbacks would needlessly complicate the main counter class. Also I realized the `value` field in the struct is completely unnecessary, since as I noted in the code, it is conveniently always the same as the low bit of `count`.

KevinOConnor commented:
Thanks.  I committed this change and made a couple of minor changes on top (commit 4d3d25b1).

-Kevin

Markus-1984 commented:
Hi,

great work with this feature. Is it possible to observe the fan rpm and start an event if the rpm is under a specified value for a specified time? In my opinion it would be very useful to for example pause the print and park the extruder if the hotend fan will stop working to avoid clogging the heatbreak.

best regards,
Markus

arkeet commented:
@Markus-1984 Klipper currently doesn't have a built-in way to take action based on the fan rpm, but it is possible to watch it with `delayed_gcode` macros. Here is an example, courtesy of @alchemyEngine:
```ini
[delayed_gcode CHECK_ALL_FANS]
initial_duration: 1
gcode:
    UPDATE_DELAYED_GCODE ID=CHECK_ALL_FANS DURATION=3
    HOTEND_FAN_CHECK

[gcode_macro HOTEND_FAN_CHECK]
variable_he_stop_count: 0
gcode:
    {% set min_rpm = 3000|float %}
    {% set max_consecutive_stops = 3 %}
    {% set rpm = printer['heater_fan hotend_fan'].rpm|float %}
    {% set he_target = printer[printer.toolhead.extruder].target|float %}
    {% set he_temp = printer[printer.toolhead.extruder].temperature|float %}
    {% set fan_on_temp = printer.configfile.settings['heater_fan hotend_fan'].heater_temp|float %}
    {% set he_stop_count = printer["gcode_macro HOTEND_FAN_CHECK"].he_stop_count|int %}

    {% if (he_target >= fan_on_temp) and (rpm < min_rpm) and (he_temp >= fan_on_temp) %}
        SET_GCODE_VARIABLE MACRO=HOTEND_FAN_CHECK VARIABLE=he_stop_count VALUE={he_stop_count + 1}
        M118 WARNING: Fan stoppage detected ({he_stop_count+1}/{max_consecutive_stops}).
        M400
        {% if printer["gcode_macro HOTEND_FAN_CHECK"].he_stop_count|int >= max_consecutive_stops-1 %}
            FAN_STOPPAGE_ROUTINE
        {% endif %}
    {% else %}
        SET_GCODE_VARIABLE MACRO=HOTEND_FAN_CHECK VARIABLE=he_stop_count VALUE=0
    {% endif %}

[gcode_macro FAN_STOPPAGE_ROUTINE]
gcode:
    M117 !!FAN BLOCKED!!
    M118 FAN STOPPAGE DETECTED. PAUSING...
    SET_HEATER_TEMPERATURE HEATER=extruder TARGET=0
    PAUSE```

Markus-1984 commented:
@arkeet 
Thanks for your quick reply.
Your solution works well!
I just have to remove the two !! in front of FAN BLOCKED!! and the ``` behind PAUSE in the FAN_STOPPAGE_ROUTINE
